### ISSUE TITLE
Packaging

### ISSUE THREAD

**Issue by mdboom** [pyodide/pyodide#38]
Adding a new package to pyodide involves editing many disjointed files.

Instead, each package should be defined as a set of independent "recipes" that the build system could pick up and build everything needed, including the package dependency graph.

`conda` comes pretty close to what is needed here, and could be adapted to support emscripten.  However, it could be overkill.  Research should be done.

**Comment by rth** [comment]
Another point is, if I understand correctly, currently Circle CI builds all packages: wouldn't that be problematic as more large packages are added (e.g. scipy)?

**Comment by mdboom** [comment]
@rth: That's absolutely correct.  We'd need a better CI setup so packages can be built independently.

**Comment by rth** [comment]
(Responding to some comments from https://github.com/iodide-project/pyodide/pull/95 to keep the discussion about packaging in one place)

From https://github.com/iodide-project/pyodide/pull/95#issuecomment-408174659

> Automatically extracting the dependencies from a pip package is basically an unsolvable problem -- even conda-build punted on it and requires that they are listed explicitly in the meta.yaml

Agreed, but pip still can still determine what to install while running setup.py, even if it's nota full dependency resolution. Would it be too far fetched to monkeypatch pip/setuptools in some why to make it also run the generation of WASM files? Or do we need to know the actual dependencies also when importing packages in the WASM environment?

From https://github.com/iodide-project/pyodide/pull/95#issuecomment-408226050

> This reminds me: I had an idea a while ago to make a tool that would create a meta.yaml from a PyPI entry, and prepopulate it with the latest version, the url and sha256 hash etc. You'd still have to manually set up the requirements, but it would go a long way to simplifying the creation of pure Python packages.

Yes, there is also the [PyPi Json API](https://warehouse.readthedocs.io/api-reference/json/#get--pypi--project_name--json) from which the latest package version, url and sha256 could determined. Though it we want dependencies, I wonder if we shouldn't rather parse the all the `meta.yaml` from conda-forge.

**Comment by rth** [comment]
A few more thoughts on this. I imagine there could be several directions to explore.

First could be to adapt some of the conda workflow as mentioned above. The `meta.yaml` appear to be already somewhat inspired by it, so I think it might make sense to grab some of the `meta.yaml` with the list of dependencies etc from conda-forge, particularly for pure python packages. If I understand correctly, we won't need to support builds under different OS, and Python versions (which is one of the main selling point of conda)? Looking at the amount of maintenance happening on conda forge, it doesn't seem that in this case a fully conda like approach could be achievable (or needed) though.  Some high level perspective on this from someone from conda-forge might be helpful.


I'm not fully acquainted with the build process yes, so this might not be technically possible. I was wondering if patching pip/setuptools to also build WASM files and e.g. bundling them somewhere in wheels would make any sense (assuming one can get around needing an explicit dependency graph). That would allow re-using pip's wheel caching mechanism, and also greatly simplify building of user defined packages (without needing to explicitly add them with a PR here). A last step could be to extract wasm artifacts from built wheels and deploying them somewhere.

For instance, currently PyPy works mostly out of the box with pip, and this greatly simplifies building packages by users. Granted, is not completely valid, but I'm still wondering if some of the issues and solutions found by PyPy regarding packaging would be relevant to pyiodide. 

Another point to consider, is that although one could package some of the main packages with a conda-like `meta.yaml` setup, in the end users would also typically need installing their custom / proprietary package mostly in pure python (typically done with pip/setuptool in a conda environment). Setting up the current build toolchain on Unix system is not something that many users might want to do, not sure what could be a way to facilitate this.

**Comment by mdboom** [comment]
This is really helpful and interesting.

We definitely need to know the package dependencies at run time.  That's how `pyodide.loadPackage('matplotlib')` is able to also load Numpy, kiwisolver etc.  Right now that's dumped into a generated `packages.json` file which is loaded in the browser.

I've spent a little time thinking about maybe using `conda-build` as the build tool for pyodide.  That is, just as Windows, Linux and MacOS are targets for `conda-build`, WASM would be a new target.  Then we can reuse its infrastructure for building packages, and the matrix of compatible versions, etc.  I think basing things on `conda` is preferable `pip/setuptools` mainly because it deals with non-Python dependencies much better, and it also handles version dependencies properly -- i.e. makes sure Numpy-dependent packages are built for the correct Numpy, and is able to pin specific versions of dependencies etc.  I think eventually we'll want to handle making multiple versions of packages available, and `conda` is the only tool I'm aware of that can handle that well for compiled packages.  As you can see, what's here is shamelessly inspired by `conda` to make an eventual transition to real `conda` hopefully a little easier down the road.  But it's quite possible that all that may not actually use `conda` on the client -- just use `conda-build` in the backend.

But, of course, the `pip/setuptools/PyPI` ecosystem is important, and we probably want a way to handle the easy case of supporting pure Python packages on PyPI in some semi-automated way.  Even the conda folks admit this and have made it easier to use `conda` with `pip` over the years.

I think grabbing existing `meta.yaml` files from conda-forge as a basis for pyodide packages seems like an efficient way to add new packages for now.  We could add a tool to remove the keys in the `meta.yaml` that pyodide doesn't understand if that proves to be time-consuming, or maybe that doesn't matter too much.

I don't know enough about wheels to know if reusing `pip/setuptools` would be helpful here.  They still don't handle non-Python dependencies well, as far as I know, so we'd have to reinvent what `conda` already solves there.  But if you have some thoughts about how that might work, by all means investigate.  I'm by no means certain of the best way forward.

As for installing custom/local libraries to the build -- that's an important use case to support.  As you've alluded to, we should probably make the build tools into an installable library.  Then we could imagine a command line tool that runs `setup.py` on a given package to create a `pyodide package`.  Should be easy for pure Python.  The tricky bit is that the `packages.json` is currently monolithic, and we probably don't want to require these custom packages to modify it.  The advantage of that approach, is that `loadPackage` can figure out all of the dependencies up front and load them in parallel -- i.e., it doesn't have to load a package first before determining its dependencies and loading those, which is unfortunately the `pip/setuptools` model.  However, we could add dependency metadata to the pyodide package itself in its `.js` file, such that you could determine the dependencies of a custom package once loading it.

There's a lot of moving pieces here.  I'm going to start a Wiki page enumerating some of the things we'd like to be able to do with packaging that might help focus priorities and decisions etc.

**Comment by rth** [comment]
In retrospective, having a build system that rebuilds all the stack from scratch has certainly some disadvantages (support of multiple versions, backward compatibilty, scaling in the long term etc) but with the latest build optimizations, we will be able to build CPython, CLAPACK, numpy, pandas, scipy, matplotlib, scikit-learn in ~35min. It certainly not that fast, but it's managable and it is very nice to be able to change some compile flag and see how it affects all the built packages at once. It certainly allows faster iterations.

 Something like changing the [`NUM_PARAMS` in emsdk](https://github.com/iodide-project/pyodide/blob/b50c8d622fc8b1730f0bb965c4eb2b271e3869a9/emsdk/patches/num_params.patch#L10) (which breaks package compatibilty) would be very annoying with a distibuted conda-forge like build systems. So maybe we should postpone changing the current build approach until everything gets stabilized a bit more and we are sure we won't need to change things in the bottom of the stack that would require a full rebuild.

**Comment by mdboom** [comment]
I agree 100%.  I think conda is the obvious long-term choice if this is to become a widely-used and distributed thing.  But while we're still in the "solving hard problem in the core" phase, it's been helpful to be monolithic.  I feel like Scipy is probably one of the hardest things to handle in the ecosystem, though, so once it's solved, hopefully required changes will settle down.

**Comment by rth** [comment]
> I feel like Scipy is probably one of the hardest things to handle in the ecosystem, though, so once it's solved, hopefully, required changes will settle down.

True, assuming no one tries to package [geopandas](https://github.com/geopandas/geopandas) that requires [DGAL](https://github.com/OSGeo/gdal) or some other GIS packages with lots of C/C++ code...

**Comment by rth** [comment]
To come back to this discussion I thin taking an inspiration from conda is very useful (and conda-forge is great) however it has following limitations in the context of web,
 - conda is so largely used, in part also because you can pip install packages from PyPi into a conda env (that are missing from conda) it and it will work. All packages will never be on conda. So that means that one has to maintain not one but two package managers. That works for conda, as they have enough resources (and pip is maintained by another org). In our case the question is what build approach could be sufficient for pyodide and work with least amount of effort. We also already micropip (and need to support it in any case).
 - I think the idea of solving dependencies on page load is not realistic for the web. Dependency resolution takes time in conda, there is https://github.com/QuantStack/mamba but it's still a large amount of effort. Instead pre-computing the dependency tree and storing it in a lock file, as Node or Poetry are doing, would be more efficient IMO. That also means that the micropip installer won't need to do anything too fancy.
 - in the current state we don't have the resource to maintain a package index, and having things more decentralized might help. Integrating into conda-forge could also help, but I'm not sure we are stable enough at present for it to happen. 

I have opened a few issues to explore related topics,
 - on building binary wheels https://github.com/iodide-project/pyodide/issues/655
 - having a pre-built docker image to facilitate building packages https://github.com/iodide-project/pyodide/issues/656
 - using `pyproject.toml` as an alternate (and optional) location for build configuration.

**Comment by henryiii** [comment]
Have you considered the cibuildwheel model? The conda-forge model requires massive centralized CI, while the cibuildwheel model provides easy tooling for independent CI. In fact, with the docker image, it might be very close to possible to add this to cibuildwheel...

We handle configuration in cibuildwheel by looking in pyproject.toml, but allowing `--config-file` to point at another file, and the structure is exactly the same.

**Comment by rth** [comment]
I am very much looking forward to making our build system modular enough to work with cibuildwheel (which is a great project BTW!). For relatively simple packages with C extensions that should be feasible. For pure Python packages (universal) wheels are already built ideally on PyPi.
The issues begin when building more complex packages (which was the initial motivation for creating conda probably). Even something like numpy, scipy bundling BLAS repeatedly inside each wheel as it's done usually is not a great idea IMO in the WASM context as download size is critical. So now we have numpy and scipy wheels but that will also load at runtime `CLAPACK-3.2.1.zip` with the LAPACK .so files, which is more similar to what conda does.

Another issue is that in an ideal world you just run "build <package name>" on the official release and be done. That might happen in the future , and should generally works for packages using e.g. Cython, other packages require a lot of patching to make them work for WASM (see e.g. [numpy](https://github.com/pyodide/pyodide/tree/main/packages/numpy/patches)). So a configuration file alone is not going to be sufficient there.

We are a bit walking the line between conda like approach and pip/PyPi like approach both for the build and for dependency installation currently.

But anyway trying to see what would be needed to make things work with cibuildwheel would be a interesting project. If you are planning to explore, we would be happy to hear your feedback. And the build system is still WIP so things can be still adjusted without too many backward compatibility constraints. xref https://github.com/pypa/cibuildwheel/issues/1002

**Comment by joerick** [comment]
From a cibuildwheel perspective (and without understanding any complexity around compilation), I can definitely see value in projects being able to build their own WASM wheels, especially to tackle the 'long-tail' of Python packages. I do see your point around WASM binary size and shared libraries too. Maybe a tiered system could work - `install numpy` would first look at the monolithic/conda-style compilation, before falling back to PyPI wheels. Probably this is how many people use conda today.

A thought on WASM PyPI wheels -

Being that WASM is a lightweight VM that can be decoupled from the browser, I can even imagine that WASM wheels might one day also be useful on the desktop side as a fallback for native code since the WASM VM is quite easy to spin up and available on a lot of platforms, and still quite a lot faster than interpreted Python code. In this way I see WASM as a general-purpose cross-platform VM, kinda like the Java VM, but small and open ;).

(The project [wasmer-python](https://github.com/wasmerio/wasmer-python) already lets you run WASM modules from within Python. note- WASM modules, not Python modules, so the APIs are not Python APIs.)

A general-purpose WASM wheel would also give package maintainers more reason to build WASM wheels.

However, this would put quite a lot of pressure on getting the WASM wheel structure/interface right. So perhaps not the right tactic for now. But it might be something to aim for as a best case a few years down the line? 

just a thought!

**Comment by ryanking13** [comment]
Closing as Pyodide packaging has changed a lot sine the last comment.

We now have out-of-build package build support that is integrated into cibuildwheel and many scientific python packages. Recipe support is also quite stable. Closing as resolved. If you have any packaging related issues, please open a new issue with specific topic.
